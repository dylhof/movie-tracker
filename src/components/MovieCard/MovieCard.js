import React from 'react'

export const MovieCard = ({title, poster_path}) => {
  const poster = `https://image.tmdb.org/t/p/w200/${poster_path}`
  const alt = `${title} poster`
  return(
    <div>
          <h2>{title}</h2>
          <img src={poster} alt={alt} />
        </div>
  )
}

// If the user is logged in
  // check if the movie is a favorite
    // yes=> display with favorite indicated
// this will display a single movie 
// 