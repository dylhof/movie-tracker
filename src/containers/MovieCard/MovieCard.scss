

.favorite-btn {
  border: 0;
  padding: 0;

  margin: 10px 50px 0 0;
  z-index: 1;
  display: block;
  align-self: flex-end;
  font-size: 32px;
  cursor: pointer;
  transition: .2s linear;
  color: white;
  background-color: transparent;
    width: 50px;
  height: 50px;
}

.isFavorite {
  color: yellow;
}

.fa-star {
  pointer-events: none;
}