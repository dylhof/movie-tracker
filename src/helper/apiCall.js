export const fetchData = async (url) => {
  const response = await fetch(url)
  if(response.ok) {
    return response.json() 
  } else {
    throw new Error(`Error fetching, code: ${response.status}`)
  }
}

export const fetchPost = async (url, options) => {
  const response = await fetch(url, options)
  if (response.ok) {
    return response.json()
  } else {
    throw new Error(response.status)
  }
}