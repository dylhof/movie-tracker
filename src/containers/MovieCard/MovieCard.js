import React, { Component } from 'react'
import { connect } from 'react-redux'
import { fetchPost } from '../../helper/apiCall'
import { addFavorite, deleteFavorite } from '../../actions';

export class MovieCard extends Component {
  constructor() {
    super()
    this.state = {
      isUser: '',
      error: ''
    }
  }

  handleFavoriteClick = async (event) => {
    const { title, id, poster_path, release_date, vote_average, overview } = this.props
    
    if (this.props.currentUser) {
      const {userID} = this.props.currentUser

      if (event.target.value === 'true') {
        try {
          const url = `http://localhost:3000/api/users/${userID}/favorites/${id}`
          const response = await fetchPost(url,
            {
              method: 'DELETE',
              body: JSON.stringify({
                user_id: userID,
                movie_id: id
              }),
              headers: {
                'Content-Type': 'application/json'
              }
            })
          this.props.dispatchDeleteFavorite(id)
        } catch (error) {
          if (error.message === '500') {
            this.setState({ error: 'Something went wrong, favorite not added' })
          }
        }
      } else {
        try {
          const response = await fetchPost('http://localhost:3000/api/users/favorites/new',
            {
              method: 'POST',
              body: JSON.stringify({
                movie_id: id,
                user_id: userID,
                title,
                poster_path,
                release_date,
                vote_average,
                overview
              }),
              headers: {
                'Content-Type': 'application/json'
              }
            })
          this.props.dispatchAddFavorite(id)
        } catch (error) {
          if (error.message === '500') {
            this.setState({ error: 'Something went wrong, favorite not added' })
          }
        }
      }
    } else {
      this.setState({
        isUser: 'Please login or create an acount to favorite your movies!'
      })
    }

  }
  render() {

    const { title, poster_path, id, favorites } = this.props
    const cssClasses = ["favorite-btn", favorites.includes(id) ? "isFavorite" : null]
    // const {currentUser} = this.props.currentUser
    const value = favorites.includes(id) ? true : false
    const poster = `https://image.tmdb.org/t/p/w200/${poster_path}`
    const alt = `${title} poster`
    return (
      <div>
        <h2>{title}</h2>
        <img src={poster} alt={alt} />
        {/* <p>{name}</p> */}
        {/*onClick make fetch to add(api/users/favorites/new)
       send(movie_id, user_id, title, poster_path, release_date vote_average, overview),
       action = ADD_FAVORITE, DELETE_FAVORITE */}
        <button className={cssClasses.join(' ')} value={value} onClick={(event) => this.handleFavoriteClick(event)}><i className='fas fa-star'></i></button>
        <span>{this.state.isUser}</span>
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  currentUser: state.currentUser,
  favorites: state.favorites,
})

export const mapDispatchToProps = (dispatch) => ({
  dispatchAddFavorite: (id) => dispatch(addFavorite(id)),
  dispatchDeleteFavorite: (id) => dispatch(deleteFavorite(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(MovieCard)



// If the user is logged in
  // check if the movie is a favorite
    // yes=> display with favorite indicated
// this will display a single movie 
// 