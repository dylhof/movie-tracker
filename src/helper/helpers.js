import * as API from './apiCall';

export const tryUnfavorite = async (id, userID) => {

  const url = `http://localhost:3000/api/users/${userID}/favorites/${id}`
  await API.fetchPost(url,
    {
      method: 'DELETE',
      body: JSON.stringify({
        user_id: userID, movie_id: id
      }),
      headers: {
        'Content-Type': 'application/json'
      }
    })
}

export const tryFavorite = async (title, id, userID, poster_path, release_date, vote_average, overview) => {

  await API.fetchPost('http://localhost:3000/api/users/favorites/new',
    {
      method: 'POST',
      body: JSON.stringify({
        movie_id: id, user_id: userID, title, poster_path,
        release_date, vote_average, overview
      }),
      headers: {
        'Content-Type': 'application/json'
      }
    })
}

