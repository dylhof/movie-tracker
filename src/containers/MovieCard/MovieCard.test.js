import React from 'react';
import { shallow } from 'enzyme';
import { MovieCard, mapStateToProps, mapDispatchToProps } from './MovieCard'
import * as helper from '../../helper/helpers'

describe('MovieCard', () => {
  let wrapper
  beforeEach(() => {
    let currentUser = { name: 'Matt', userID: 1 }
    let mockDispatchDeleteFavorite = jest.fn()
    let mockDispatchAddFavorite = jest.fn()
    let mockDispatchSetError = jest.fn()
    let mockFavorites = [1, 2, 3]
    wrapper = shallow(
      <MovieCard
        currentUser={currentUser}
        dispatchDeleteFavorite={mockDispatchDeleteFavorite}
        dispatchAddFavorite={mockDispatchAddFavorite}
        dispatchSetError={mockDispatchSetError}
        favorites={mockFavorites}
      />
    )
  })

  it('should match the snapshot', () => {

  })

  it('should have initial state', () => {
    //setup
    const expected = {isUser: ''}
    //expectation
    expect(wrapper.state()).toEqual(expected)
  })

  describe('handleFavoriteClick', () => {

    it('should call tryUnfavorite if currentUser and already favorite', () => {
      //setup
      const mockEvent = { target: { value: 'true' } }
      helper.tryUnfavorite = jest.fn()
      //execution
      wrapper.instance().handleFavoriteClick(mockEvent)
      //expectation
      expect(wrapper.instance().props.dispatchDeleteFavorite).toHaveBeenCalled()
      expect(helper.tryUnfavorite).toHaveBeenCalled()
    })

    it('should call tryFavorite if currentUser and not already favorite', () => {
      //setup 
      const mockEvent = { target: { value: 'false' } }
      helper.tryFavorite = jest.fn()
      //execution
      wrapper.instance().handleFavoriteClick(mockEvent)
      //expectation
      expect(wrapper.instance().props.dispatchAddFavorite).toHaveBeenCalled()
      expect(helper.tryFavorite).toHaveBeenCalled()
    })

    it('should set state of isUser if there is no current user', () => {
      //setup
      const mockEvent = { target: { value: 'false' } }
      let mockFavorites = [1, 2, 3]
      let currentUser = null
      wrapper = shallow(
        <MovieCard
          currentUser={currentUser}
          favorites={mockFavorites}
        />)
      //execution
      wrapper.instance().handleFavoriteClick(mockEvent)
      //expectation
      expect(wrapper.state('isUser')).toEqual('Please login or create an acount to favorite your movies!')
    })
  })
})