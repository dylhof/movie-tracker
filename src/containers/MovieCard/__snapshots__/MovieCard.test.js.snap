// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard should match the snapshot 1`] = `
<div
  className="movie-card"
>
  <div
    className="title-favorite-div"
  >
    <h2 />
    <button
      className="favorite-btn "
      onClick={[Function]}
      value={false}
    >
      <i
        className="fas fa-star"
      />
    </button>
  </div>
  <Link
    replace={false}
    to="/movie/undefined"
  >
    <img
      alt="undefined poster"
      className="movie-poster"
      src="https://image.tmdb.org/t/p/w200/undefined"
    />
  </Link>
</div>
`;
